### DISCUSSION TITLE
Getting roles that has the same exact permissions

### DISCUSSION THREAD

**Discussion by Abdelhalim-Merabet** [spatie/laravel-permission#1976] [Q&A] [answered]
hello everyone, hope you are doing fine

in laravel 8 , am using the spatie package for managing roles and permissions , and it's doing great

I found that users can create roles with the same exact permissions (roles has many permissions) over and over again with different names

for example , "Users Manager" would have these permissions :
- users.edit
- users.create
- users.delete

however , the app admin can create another role "Boo Role" for example and give it the same exact permissions

so my question is, how can i check if any existing role has the same exact given permissions by the user before creating a new one

**Comment by erikn69** [comment, marked as answer]
Get roles with a `whereHas` and in the whereHas add a `whereIn` with the Permission ids, and add a `withCount` for the exact number of Permission, and compare

**Reply by Abdelhalim-Merabet** [reply]
```php
$roles = Role::whereHas('permissions', function ($query) use ($ids) {
                $query->whereIn('id', $ids);
            }, '=', count($ids))->get();
```
like this ? this would parse roles if they has more permissions

**Reply by erikn69** [reply]
It is ok if works, I mean `withCount`, but there is many ways

**Reply by Abdelhalim-Merabet** [reply]
it doesn't work , it returns roles that have either Permission "1" or "2" , i want them to be exactly "1" and "2"

**Reply by erikn69** [reply]
```php
$similar_role = Role::withCount('permissions')
    ->whereHas('permissions', fn($query) => $query->whereIn('id', $selectedPermissions))
    ->having('permissions_count', count($selectedPermissions))->get();
```

**Comment by Abdelhalim-Merabet** [comment]
this is the current solution that i got ; putting it here maybe it ll help some one
this will return roles that have either Permission "1" or "2" ( my question was about getting exactly just "1" and "2")
```php
$similar_role = Role::withCount('permissions')->whereHas('permissions', function ($query) use ($selectedPermissions) {
            $query->whereIn('id', $selectedPermissions);
        }, '=', count($selectedPermissions))->get();
```
then check if the count equals the selected new permissions to verify that the Role already exist
```php
        $similar_role = $similar_role->filter(function ($item) use ($selectedPermissions) {
            return $item->permissions_count === count($selectedPermissions);
        });
```

**Reply by erikn69** [reply]
The same answer that I give you with the `withCount` as i say, but you take the credit 🙄

**Reply by Abdelhalim-Merabet** [reply]
sorry didn't understand ! u meant a second line for checking the count ? 
i thought u meant inside the eloquent query , wich didn't work
didn't mean any disrespect 😅 !

**Reply by Abdelhalim-Merabet** [reply]
don't you think that a method that solves this should be in the package ?

**Reply by erikn69** [reply]
Not for the package, also you can use `having('permissions_count',count($selectedPermissions))` or `get()->where('permisions_count', count($selectedPermissions)) ->count()>0` instead of filter, Laravel collection has many methods, and you can delete ` '=', count($selectedPermissions))` on the query

```php
$similar_role = Role::withCount('permissions')
    ->whereHas('permissions', fn($query) => $query->whereIn('id', $selectedPermissions))
    ->having('permissions_count', count($selectedPermissions))->get();
```
